pytorch_lightning/core/optimizer.py

```diff
@@ -79,6 +102,31 @@ def _should_accumulate(self):
         is_final_batch = self._trainer.train_loop._num_training_batches_reached()
         return not (accumulation_done or is_final_batch)
 
+    def __optimizer_step(self, *args, closure: Optional[Callable] = None, profiler_name: str = None, **kwargs):
+        trainer = self._trainer
+        optimizer = self._optimizer
+        model = trainer.get_model()
+
+        if trainer.on_tpu:
+            with trainer.profiler.profile(profiler_name):
+                xm.optimizer_step(optimizer, optimizer_args={'closure': closure, **kwargs})
+
+        elif trainer.amp_backend is not None:
+            trainer.precision_connector.backend.optimizer_step(trainer, optimizer, closure)
+
+        else:
+            with trainer.profiler.profile(profiler_name):
+                optimizer.step(closure=closure, *args, **kwargs)
+
+        trainer.train_loop.on_before_zero_grad(self)
+
+        model.optimizer_zero_grad(
+            trainer.current_epoch,
+            trainer.batch_idx,
+            optimizer,
+            self._optimizer_idx
+        )
```

Comment:
it won't be called if `enable_pl_optimizer=False`??

Reply:
When people are using `automatic_optimization=True` and `enable_pl_optimizer=False` without overriding `optimizer_step`.

We will wrap their optimizer only the time to run step. So this function zero_grad will be called.
However, if they override `optimizer_step` with  Trainer(automatic_optimization=True, enable_pl_optimizer=False), they will get a warning we can't take care of zero_grad for them.

```
        if not isinstance(optimizer, LightningOptimizer):
            # wraps into LightingOptimizer only for running step
            optimizer = LightningOptimizer.to_lightning_optimizer(optimizer, self.trainer)
        optimizer.step(closure=optimizer_closure, *args, **kwargs)
```